performance_test/src/communication_abstractions/opendds_communicator.hpp

```diff
@@ -0,0 +1,344 @@
+
+      dw_qos.resource_limits.max_samples = 32;
+      dw_qos.resource_limits.max_samples_per_instance = 32;
+      dw_qos.resource_limits.max_instances = 1;
+
+      OpenDdsQOSAdapter qos_adapter(m_ec.qos());
+      qos_adapter.apply_dw(dw_qos);
+
+      m_datawriter = publisher->create_datawriter(m_topic,
+          dw_qos, nullptr, OpenDDS::DCPS::DEFAULT_STATUS_MASK);
+      if (CORBA::is_nil(m_datawriter)) {
+        throw std::runtime_error("Could not create datawriter");
+      }
+
+      m_typed_datawriter = DataWriterType::_narrow(m_datawriter);
+      if (CORBA::is_nil(m_typed_datawriter)) {
+        throw std::runtime_error("failed datawriter narrow");
+      }
+    }
+    lock();
+    data.time_ = time.count();
+    data.id_ = next_sample_id();
+    increment_sent();  // We increment before publishing so we don't have to lock twice.
+    unlock();
+    auto retcode = m_typed_datawriter->write(data, DDS::HANDLE_NIL);
+    if (retcode != DDS::RETCODE_OK) {
+      throw std::runtime_error("Failed to write to sample");
+    }
+  }
+
+  /**
+   * \brief Reads received data from DDS.
+   *
+   * In detail this function:
+   * * Reads samples from DDS.
+   * * Verifies that the data arrived in the right order, chronologically and also consistent with the publishing order.
+   * * Counts received and lost samples.
+   * * Calculates the latency of the samples received and updates the statistics accordingly.
+   */
+
+  void update_subscription()
+  {
+    if (CORBA::is_nil(m_datareader)) {
+      DDS::Subscriber_ptr subscriber = nullptr;
+      DDS::DataReaderQos dr_qos;
+      ResourceManager::get().opendds_subscriber(subscriber, dr_qos);
+
+      dr_qos.resource_limits.max_samples = 32;
+      dr_qos.resource_limits.max_instances = 10;
+      dr_qos.resource_limits.max_samples_per_instance = 32;
+      /* if there are more remote writers, you need to increase these limits */
+      
+      //These are not available in OpenDDS, RTI tricks
+      //dr_qos.reader_resource_limits.max_remote_writers = 10;
+      //dr_qos.reader_resource_limits.max_remote_writers_per_instance = 10;
+
+      OpenDdsQOSAdapter qos_adapter(m_ec.qos());
+      qos_adapter.apply_dr(dr_qos);
+
+      /* Only DDS_DATA_AVAILABLE_STATUS supported currently */
+      m_datareader = subscriber->create_datareader(
+        m_topic,
+        dr_qos,
+        nullptr,
+        OpenDDS::DCPS::DEFAULT_STATUS_MASK);
+
+      if (CORBA::is_nil(m_datareader)) {
+        throw std::runtime_error("datareader == nullptr");
+      }
+
+      m_condition = m_datareader->get_statuscondition();
+      m_condition->set_enabled_statuses(DDS::DATA_AVAILABLE_STATUS);
+      m_waitset.attach_condition(m_condition);
+
+      m_typed_datareader = DataReaderType::_narrow(m_datareader);
+      if (m_typed_datareader == nullptr) {
+        throw std::runtime_error("m_typed_datareader == nullptr");
+      }
+
+      //Again, RTI trick. No wonder
+      //if (!m_condition_seq.ensure_length(2, 2)) {
```

Comment:
This is just ensuring that the sequence has enough elements allocated to hold all of our conditions - the RTI sequences won't dynamically resize (I think the OpenDDS ones do if I recall correctly). If there is a `reserve()` method or similar you could do that, but it shouldn't matter too much since we will reuse the sequence so we will only have to allocate the first time through this method

Reply:
I just removed it for now